src/ros2idl.ne

```diff
@@ -0,0 +1,395 @@
+}, {});
+
+const moo = require("moo");
+// Terminal tokens are in all caps
+const lexer = moo.compile({
+  SPACE: {match: /\s+/, lineBreaks: true},
+  DECIMALEXP: /(?:(?:\d+\.\d*)|(?:\d*\.\d+)|(?:[0-9]+))[eE](?:[+|-])?[0-9]+/,
+  DECIMAL: /(?:(?:\d+\.\d*)|(?:\d*\.\d+))/,
+  INTEGER: /[0-9]+/,
+  COMMENT: /(?:\/\/[^\n]*)|(?:\/\*(?:.|\n)+?\*\/)/,
+  HEX_LITERAL: /0x(?:[0-9][a-f][A-F])+?/,
+  STRING: [{match: /"(?:\\["\\rnu]|[^"\\])*?"/, lineBreaks: true, value: x => x.slice(1, -1)}], // remove outside quotes
+  LCBR: '{',
+  RCBR: '}',
+  LBR: '[',
+  RBR: ']',
+  LT: '<',
+  GT: '>',
+  LPAR: '(',
+  RPAR: ')',
+  ';': ';',
+  ',': ',',
+  AT: '@',
+  PND: '#',
+  PT: ".",
+  '/': "/",
+  SIGN: /[+|-]/,
+  EQ: /=[^\n]*?/,
+  NAME: {match: /[a-zA-Z_][a-zA-Z0-9_]*(?:\:\:[a-zA-Z][a-zA-Z0-9_]*)*/, type: moo.keywords(kwObject)},
+});
+
+// Utiility functions
+
+// also used to parse tokens to strings since they start as an object
+function join(d){
+	return d.join("");
+}
+
+// used for combining AST components
+function extend(objs) {
+  return objs.reduce((r, p) => ({ ...r, ...p }), {});
+}
+
+function noop() {
+	return null;
+}
+
+// Constants can be used in defs to define lengths of arrays and strings
+// they should be defined before the def that uses them
+const constantToValueMap = {};
+
+function getIntOrConstantValue(d) {
+	const int = parseInt(d);
+	if(!isNaN(int)) {
+		return int
+	}
+
+	// handle %NAME value
+	return d?.value ? constantToValueMap[d.value] : undefined;	
+}
+
+function processComplexModule(d) {
+  const moduleName = d[0][4].name;
+  const defs = d[0][8];
+  // returning array of message definitions
+  const msgDefs = [];
+  function traverse(node, processNode) {
+    if(Array.isArray(node)) {
+      node.forEach(n => traverse(n, processNode));
+    } else {
+      processNode(node);
+    }
+  }
+  // Need to update the names of modules and structs to be in their respective namespaces
+  traverse(d[0], (sub) => {
+    if(sub && sub.definitions) {
+      sub.name = `${moduleName}::${sub.name}`;
+      msgDefs.push(sub);
+    }
+  });
+  
+  
+  return msgDefs;
+}
+
+function processConstantModule(d) {
+	const moduleName = d[0][4].name;
+  const enclosedConstants = d[0][8];
+  // need to return array here to keep same signature as processComplexModule
+	return [{
+		name: moduleName,
+		definitions: enclosedConstants.flat(1),
+	}];
+}
+
+%}
+
+@lexer lexer
+
+main -> (importDcl __):* module:+ _ {% d => {
+	return d[1][0];
+}
+%}
+
+# support <import> or "import" includes - just ignored
+importDcl -> _ "#" "include" __ (%STRING | "<" _ %NAME ("/" %NAME):* "." "idl" _ ">") {% noop %}
+
+# constant modules need to be separate from complex modules since they shouldn't mix
+module  -> ((comment|annotation):* _ "module" __ fieldName __ "{" __ (constantDcl):+ __ "}" semi) {% processConstantModule %}
+  | ((comment|annotation):* _ "module" __ fieldName __ "{" __ (structWithAnnotation|module):+ __ "}" semi) {% processComplexModule %}
```

Comment:
is there a reason that `comment` has to be here at the beginning? can it be up at the `main` level?

Alternatively, I suppose maybe we technically need to support comments occurring anywhere on a line (like `module /*comment*/ { ... }` ... is that something the lexer can remove so we don't have to deal with them in the grammar?

Reply:
> using regex to find comments in it will also find comments in strings

Can this be solved by making the string matcher come before the comment matcher in the lexer definition?

Reply:
Fixed it by overriding the `.next` on the lexer. comment and whitespace tokens are now entirely ignored